### PR TITLE
Migrate to Smart Cache Core Libraries for Kafka functionality (CORE-684)

### PR DESCRIPTION
**Pull request by rvesse**
This PR represents a 2.x version of these libraries.  It reworks the internals considerably to be based upon reusing the Kafka functionality encapsulated, and much more heavily tested, in our Smart Cache Core Libraries from telicent-oss/smart-caches-core.  

It is undoubtably a monster PR but it is a reduction in total lines of code, much of the added lines of code are in new/improved unit and integration tests not in product code.  The following are calculated metrics for this branch vs `main`:

Metrics for this branch
Metrics for main

If we look at `LOCp` (Lines of Product Code) as the headline metric then this PR reduces the codebase by roughly half while gaining a ton of improvements described below.

Notable changes are:

- Uses various Smart Cache `EventSource` and related APIs to simplify implementation and provide more functionality:
    - Support for multi-partition topics
    - Support for having a single connector read from multiple topics
    - Support for DLQ for malformed events
    - Ensuring that events that preceded a malformed event in a transaction batch are processed and not lost
- Lots of bug fixes around Consumer Group handling
- Improved unit and integration test coverage

# Outstanding Work

- [x] Publish necessary associated changes to Smart Cache Core Libraries
- [x] Re-implement support for restoring Kafka offsets as utilised by Smart Cache Graph backup/restore functionality
- [x] Further improved unit and integration test coverage, especially around corner case and failure handling scenarios

# Next Steps

Once we're happy with this PR the next step will be to look at switching telicent-oss/smart-cache-graph over to the new `2.0.0-SNAPSHOT` builds of this to test out the changes in an end to end setup, and address any functional regressions that may have inadvertently been introduced by these changes.  Once we're happy this change is working as intended there, then we can do a `2.0.0` release and make the switchover.

### CODE REVIEW

File: jena-fuseki-kafka-module/src/main/java/org/apache/jena/fuseki/kafka/FMod_FusekiKafka.java [hunk 1]
```diff
@@ -84,50 +85,50 @@ public void prepare(FusekiServer.Builder builder, Set<String> names, Model confi
...
-        recordConnector(builder, conn, dataState);
+        String datasetName = conn.getDatasetName();
+        FusekiOffsetStore offsets = FusekiOffsetStore.builder()
+                                                     .datasetName(datasetName)
+                                                     .stateFile(conn.getStateFile() != null ?
+                                                                new File(conn.getStateFile()) : null)
+                                                     .consumerGroup(conn.getConsumerGroupId())
+                                                     .build();
+        for (Map.Entry<String, Object> offset : offsets.offsets()) {
+            FmtLog.info(LOG, "Initial offset for topic partition %s = %s (%s)", offset.getKey(), offset.getValue(),
```
Comment (rvesse) [inline jena-fuseki-kafka-module/src/main/java/org/apache/jena/fuseki/kafka/FMod_FusekiKafka.java:118]: - [ ] Add `[{}]` topic names prefix to align with other log messages

File: jena-fuseki-kafka-module/src/main/java/org/apache/jena/fuseki/kafka/FKS.java [hunk 2]
```diff
@@ -16,411 +16,269 @@
...
-import org.apache.jena.kafka.common.DataState;
+import org.apache.jena.kafka.common.FusekiOffsetStore;
+import org.apache.jena.kafka.common.FusekiProjector;
+import org.apache.jena.kafka.common.FusekiSink;
 import org.apache.jena.sparql.core.DatasetGraph;
-import org.apache.kafka.clients.NetworkClient;
-import org.apache.kafka.clients.consumer.Consumer;
-import org.apache.kafka.clients.consumer.KafkaConsumer;
-import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
```
Comment (TelicentPaul) [inline jena-fuseki-kafka-module/src/main/java/org/apache/jena/fuseki/kafka/FKS.java:45]:
```suggestion
```
Utterly pedantic - unused import

File: jena-kafka-connector/src/main/java/org/apache/jena/kafka/KConnectorDesc.java [hunk 3]
```diff
@@ -16,31 +16,42 @@
 
 package org.apache.jena.kafka;
 
+import java.util.Collections;
```
Comment (TelicentPaul) [inline jena-kafka-connector/src/main/java/org/apache/jena/kafka/KConnectorDesc.java:19]:
```suggestion
```
Utterly pedantic - unused import

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,80 @@
...
+
+- Introduced Telicent Smart Cache Core libraries to handle all the Kafka read/write logic.  These APIs provide wrappers
+  around the lower level Kafka APIs that make it easier for us to handle errors, avoid lost events etc.
+- Kafka Connectors can now be optionally configured with a Dead Letter Queue (DLQ) to which bad events are forwarded
+    - If no DLQ is configured the Kafka Connector will abort
+- Kafka Connectors can now be configured with multiple topics so a single connector can read events from multiple topics
+  into a single dataset
+    - Previously this required configuring a connector for each unique topic
+- Kafka Connectors now support reading events from multi-partition topics
+    - Previously they would only read from partition `0` irregardless of how many partitions the topic had
```
Comment (TelicentPaul) [inline CHANGELOG.md:45]:
```suggestion
    - Previously they would only read from partition `0` regardless of how many partitions the topic had
```
Utterly pedantic - fixing typo

File: jena-kafka-connector/src/main/java/org/apache/jena/kafka/KafkaConnectorAssembler.java [hunk 5]
```diff
@@ -83,57 +81,61 @@ public class KafkaConnectorAssembler extends AssemblerBase implements Assembler
...
-    public static String getNS() { return NS; }
 
-    /** Type of a connector description */
-    public static Resource tKafkaConnector = ResourceFactory.createResource(NS+"Connector");
+    public static String getNS() {
+        return NS;
+    }
+
+    /**
+     * Type of a connector description
```
Comment (TelicentPaul) [inline jena-kafka-connector/src/main/java/org/apache/jena/kafka/KafkaConnectorAssembler.java:90]:
```suggestion
     * Type of connector description
```
Utterly pedantic - fixing grammatical error

**Review by TelicentPaul — APPROVED** [review]
Looks ok to my eyes - I've added so utterly pedantic suggestions.

The only thing that is a concern is that I had failing tests on every run. They worked (on intellij) on re-run. 

[ERROR] Failures:
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenConsumingFromTopic_thenExpectedMessagesFound:125 expected object to not be null
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded:151 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded:165 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO]
[ERROR] Tests run: 44, Failures: 3, Errors: 0, Skipped: 0

I'd recommend adding 
```
<rerunFailingTestsCount>1</rerunFailingTestsCount> 
```
to the surefire plugin in order to give us more leverage.

**Comment by rvesse** [comment]
@TelicentPaul Tweaked the tests slightly including the suggested surefire plugin configuration change, no idea if that will make the tests more stable on your system or not as they work for me

**Comment by TelicentPaul** [comment]
@rvesse I'll run them again now. 

To split the toin coss, how does it run for you @robwtelicent ?

**Comment by TelicentPaul** [comment]
Alas, even with the re-runs.

[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenConsumingFromTopic_thenExpectedMessagesFound:125 expected object to not be null
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded:151 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[ERROR]   DockerTestSecureKafka>DockerTestFK.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded:165 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO]
[ERROR] Tests run: 44, Failures: 3, Errors: 0, Skipped: 0

If it's working for you, I'll drop it as it was running individually in Intellij ok

File: jena-kafka-connector/src/main/java/org/apache/jena/kafka/common/FusekiProjector.java [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+ * to process all 100 events prior to committing the transaction.
+ * </p>
+ * <p>
+ * It will also use metadata from the event source to decide when to commit the batch.  Again supposed you set the batch
+ * size to 25 but there are fewer than 25 events available, once the event source reports that it has no events
+ * remaining then the projector guarantees to commit the batch.
+ * </p>
+ * <p>
+ * Note that in pathological cases - a slow upstream producer writing to the consumed topic - this could result in
+ * single event transactions which can have adverse event on database disk size when using TDB2.  However, this scenario
```
Comment (robwtelicent) [inline jena-kafka-connector/src/main/java/org/apache/jena/kafka/common/FusekiProjector.java:54]: I think you mean `adverse effect`?

File: jena-kafka-connector/src/main/java/org/apache/jena/kafka/common/FusekiProjector.java [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+    /**
+     * Materialises the values of the event, a {@link RdfPayloadException} will be thrown if the event has a malformed
+     * payload
+     *
+     * @param event Event
+     */
+    protected final void materialiseValue(Event<Bytes, RdfPayload> event) {
+        // Try to materialize the payload value, this either succeeds if the payload is valid or throws an exception
+        // if the payload is malformed, we handle malformed payloads in our catch block
```
Comment (robwtelicent) [inline jena-kafka-connector/src/main/java/org/apache/jena/kafka/common/FusekiProjector.java:275]: Repeats (almost) the method documentation.

**Comment by rvesse** [comment]
@TelicentPaul Further modified the test setup so each test should have a fully isolated Kafka cluster which in my local testing seems to resolve the intermittent failures of some of the tests, and is now passing again on the build server.  Can you test these out on your system and make sure they work for you too?

**Review by robwtelicent — COMMENTED** [review]
A few minor comments. Always good to see a reduction in the size of the codebase!

File: jena-kafka-connector/src/main/java/org/apache/jena/kafka/KafkaConnectorAssembler.java [hunk 8]
```diff
@@ -280,86 +287,94 @@ private static void loadKafkaPropertiesFile(Node node, Properties props, File fi
...
+                SELECT ?n {
+                  OPTIONAL { ?X ?fusekiServiceName ?N1 }
+                  OPTIONAL { ?X ?fusekiDatasetName ?N2 } # Old name.
+                  BIND(COALESCE( ?N1, ?N2, '' ) AS ?n)
+                }
+                """;
+        try (QueryExec exec = QueryExec.graph(graph).query(queryString)
+                                       .substitution("X", node)
+                                       .substitution("fusekiServiceName", pFusekiServiceName)
+                                       .substitution("fusekiDatasetName", pFusekiDatasetName)
```
[the remark this answers is not in the document]
Comment (rvesse) [reply jena-kafka-connector/src/main/java/org/apache/jena/kafka/KafkaConnectorAssembler.java:312]: Think it was already replaced by `pFusekiServiceName` according to the Javadoc, since this is a major version bump I will just go ahead and delete it

**Comment by rvesse** [comment]
> A few minor comments. Always good to see a reduction in the size of the codebase!

Thanks, have done some further cleanup based on your comments

**Comment by rvesse** [comment]
Yeah I think my changes this morning may finally have cracked the test stability problem 🤞

**Comment by robwtelicent** [comment]
> Only two failures for me but if it's working for everyone else and the build, I'm fine.
> 
> [ERROR] Failures: [ERROR] DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded:116 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return but it returned for input of <0> within 15 seconds. [ERROR] DockerTestSecureKafka>DockerTestFK.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded:130 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return but it returned for input of <0> within 15 seconds. [INFO] [ERROR] Tests run: 41, Failures: 2, Errors: 0, Skipped: 0

If it is a timeout problem is it possible to increase the timeout period?

**Comment by rvesse** [comment]
@robwtelicent @TelicentPaul I've made a further commit that changes how tests are retried using TestNG retry analyzers that we use in other repositories and has greatly reduced Kafka test instability in those.  If you could both try building locally with the latest commit and see if that gets things building stably for you?

**Comment by robwtelicent** [comment]
> @robwtelicent @TelicentPaul I've made a further commit that changes how tests are retried using TestNG retry analyzers that we use in other repositories and has greatly reduced Kafka test instability in those. If you could both try building locally with the latest commit and see if that gets things building stably for you?

Builds ok for me.

**Comment by TelicentPaul** [comment]
I kicked off a run but it is taking an *age* so I think it's gone wrong. 

Will kill it and try again.

**Comment by TelicentPaul** [comment]
[ERROR] Failures:
[ERROR] org.apache.jena.fuseki.kafka.DockerTestSecureKafka.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded
[INFO]   Run 1: PASS
[INFO]   Run 2: PASS
[INFO]   Run 3: PASS
[ERROR]   Run 4: DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded:117 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO]
[ERROR] org.apache.jena.fuseki.kafka.DockerTestSecureKafka.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded
[INFO]   Run 1: PASS
[INFO]   Run 2: PASS
[INFO]   Run 3: PASS
[ERROR]   Run 4: DockerTestSecureKafka>DockerTestFK.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded:131 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO]
[INFO]
[ERROR] Tests run: 41, Failures: 2, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Jena Fuseki-Kafka Connector 2.0.0-SNAPSHOT:
[INFO]
[INFO] Apache Jena Fuseki-Kafka Connector ................. SUCCESS [  0.294 s]
[INFO] Jena-Kafka : Connector ............................. SUCCESS [02:04 min]
[INFO] Jena-Kafka : Fuseki-Kafka Connector ................ FAILURE [28:45 min]
[INFO] Jena-Kafka : Fuseki Module ......................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  30:51 min
[INFO] Finished at: 2025-02-10T15:16:51Z
[INFO] ------------------------------------------------------------------------

**Comment by TelicentPaul** [comment]
Screenshot 2025-02-12 at 14 46 45
Not much to debug even when running in Intellij - 15 minutes in before I cancelled it.

**Comment by rvesse** [comment]
> Screenshot 2025-02-12 at 14 46 45 Not much to debug even when running in Intellij - 15 minutes in before I cancelled it.

Annoyingly I ran the build in a loop 20 times on the trot yesterday while working on other things without a single failure so there's clearly some unseen/unaccounted for factor that's causing the instability on your system.

Given we have a regular release coming up I've marked this as **DO NOT MERGE** for the time being and we can pick this up again after the release closes out

**Comment by rvesse** [comment]
@TelicentPaul Could I get you to try the latest version of this PR yet again?

I know in a discussion on another issue you mentioned you'd had some Docker issues with running out of disk space that you'd since cleaned up so wonder if that was contributing to the previous failures on your system?

**Comment by TelicentPaul** [comment]
Doesn't look good, although I've just discovered - almost catastrophically; that my machine had *no* disk space. I've had to restart my laptop and clear out a huge amount of stuff. 

Hopefully, that frees up resources for the tests to run correctly. The failures are related to it not returning in X seconds.

**Comment by rvesse** [comment]
Got some logs off @TelicentPaul to see what's happening on his system.  It specifically seems to be `DockerTestSecureKafka` that is having issues:

```
[ERROR] Failures: 
[ERROR] org.apache.jena.fuseki.kafka.DockerTestSecureKafka.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded
[INFO]   Run 1: PASS
[INFO]   Run 2: PASS
[INFO]   Run 3: PASS
[ERROR]   Run 4: DockerTestSecureKafka>DockerTestFK.givenBlankOffsets_whenRunningFusekiKafka_thenDataIsLoaded:118 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO] 
[ERROR] org.apache.jena.fuseki.kafka.DockerTestSecureKafka.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded
[INFO]   Run 1: PASS
[INFO]   Run 2: PASS
[INFO]   Run 3: PASS
[ERROR]   Run 4: DockerTestSecureKafka>DockerTestFK.givenOffsetsRestored_whenRunningFusekiKafka_thenDataIsLoaded:132 » ConditionTimeout Lambda expression in org.apache.jena.fuseki.kafka.DockerTestConfigFK expected the predicate to return <true> but it returned <false> for input of <0> within 15 seconds.
[INFO] 
```

And digging into the logs further we see the following:

```
org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: fetchMetadata
2025-03-04 11:43:46,268 [Server] [<anon>] INFO  AdminMetadataManager - [AdminClient clientId=adminclient-55] Metadata update failed
org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: fetchMetadata
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Topic tests not present in metadata after 60000 ms.
	at org.apache.kafka.clients.producer.KafkaProducer$FutureFailure.<init>(KafkaProducer.java:1513)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:1115)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:993)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:875)
	at org.apache.jena.fuseki.kafka.FKLib.sendBody(FKLib.java:117)
	at org.apache.jena.fuseki.kafka.FKLib.sendFile(FKLib.java:110)
	at org.apache.jena.fuseki.kafka.FKLib.sendFiles(FKLib.java:94)
	at org.apache.jena.fuseki.kafka.FKLib.sendFiles(FKLib.java:83)
	at org.apache.jena.fuseki.kafka.DockerTestFK.setupTest(DockerTestFK.java:71)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:141)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:71)
	at org.testng.internal.invokers.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:400)
	at org.testng.internal.invokers.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:333)
	at org.testng.internal.invokers.TestInvoker.runConfigMethods(TestInvoker.java:833)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:600)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:230)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:63)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:992)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:203)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:154)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:134)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at org.testng.TestRunner.privateRun(TestRunner.java:739)
	at org.testng.TestRunner.run(TestRunner.java:614)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:421)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:413)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:373)
	at org.testng.SuiteRunner.run(SuiteRunner.java:312)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1274)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1208)
	at org.testng.TestNG.runSuites(TestNG.java:1112)
	at org.testng.TestNG.run(TestNG.java:1079)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:155)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:169)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:88)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:137)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: org.apache.kafka.common.errors.TimeoutException: Topic tests not present in metadata after 60000 ms.
2025-03-04 11:43:46,835 [Server] [<anon>] ERROR FusekiKafka - [tests] Error: sendFile src/test/files/data.nq
```

So it's failing to write the test data out to Kafka, hence there's no data for the test to read in and the test fails.  The stack trace seems to imply that the `tests` topic is not getting correctly created on the cluster.  There's unfortunately not really anything else enlightening in the logs.  May try tweaking one of the timeouts to allow more time for the topic to be created.
